### PR TITLE
#1-4Task. Vlad.Selikh.ais

### PR DESCRIPTION
**Pull request by VladSelikh**
https://travis-ci.org/github/VladSelikh/js-assignments/builds/747362754

### CODE REVIEW

**Comment by VladSelikh** [comment]
1-4 Tasks Finished.

File: task/01-strings-tasks.js [hunk 1]
```diff
@@ -22,7 +22,8 @@
  *   '',  'bb'  => 'bb'
  */
 function concatenateStrings(value1, value2) {
-    throw new Error('Not implemented');
+    let new_string=value1+value2;
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: Please, here and below try to follow these rules https://javascript.info/coding-style
(spaces around operators here)

File: task/01-strings-tasks.js [hunk 2]
```diff
@@ -99,7 +104,16 @@ function getFirstChar(value) {
  *   '\tHello, World! ' => 'Hello, World!'
  */
 function removeLeadingAndTrailingWhitespaces(value) {
-    throw new Error('Not implemented');
+    let last_i=value.length-1;
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: Redundant solution, please use trim
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim

File: task/01-strings-tasks.js [hunk 3]
```diff
@@ -114,7 +128,13 @@ function removeLeadingAndTrailingWhitespaces(value) {
...
  */
 function repeatString(value, count) {
-    throw new Error('Not implemented');
+    let i=1;
+    let new_str='';
+      while (i<=count) {
+      new_str+=value;
+      i++;
+      }
+      return new_str;
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: Please, use repeat
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/repeat

File: task/01-strings-tasks.js [hunk 4]
```diff
@@ -130,7 +150,11 @@ function repeatString(value, count) {
  *   'ABABAB','BA' => 'ABAB'
  */
 function removeFirstOccurrences(str, value) {
-    throw new Error('Not implemented');
+    let first_i=str.indexOf(value);
+    let first_str=str.substring(0, first_i);
+    let second_str=str.substring((first_i+value.length), str.length);
+    let new_str=first_str+second_str;
+    return new_str;
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: We can use String.prototype.replace
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace

File: task/01-strings-tasks.js [hunk 5]
```diff
@@ -145,8 +169,17 @@ function removeFirstOccurrences(str, value) {
...
+    let str_len=str.length-1;
+    let i=0;
+    let t=0;
+    while (i<=str_len && str[i]!==`<`) {
+    i++;
+    }
+    while (t<=str_len && str[t]!==`>`) {
+    t++;
+    }
+    return str.substring(i+1, t);
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: We can solve it just using replace '<' and '>' with ''

File: task/01-strings-tasks.js [hunk 6]
```diff
@@ -174,7 +208,20 @@ function convertToUpperCase(str) {
...
+    while (i<=str.length-1) {
+        if (str[i]===`;`) {
+        email_array[array_i]=str.substring(first_i, i);
+        first_i=i+1;
+        array_i++;
+        }
+        i++;
+    }
+    email_array[array_i]=str.substring(first_i, str.length);
+    return email_array;
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: Please, use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split

File: task/01-strings-tasks.js [hunk 7]
```diff
@@ -201,7 +248,40 @@ function extractEmails(str) {
...
+        rect_string=rect_string+`└`;
+      } else if (i===total) {
+        rect_string=rect_string+`┘\n`;
+      } else {
+        rect_string=rect_string+`─`;
+      }
+      }
+      i++;
+    }
+    return rect_string;
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: Here and below , please, don't forget about coding style

File: task/01-strings-tasks.js [hunk 8]
```diff
@@ -238,14 +331,19 @@ function encodeToRot13(str) {
  *   isString(new String('test')) => true
  */
 function isString(value) {
-    throw new Error('Not implemented');
+  if (typeof value==='string') {
+    return true;
+    } else if (typeof value==='object' && value!==null && typeof value.valueOf()==='string') {
+      return true;
+    }
+    return false;
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: Just  `return (typeof (value) === 'string') || ( value instanceof String )`
Please , read 
https://javascript.info/logical-operators
https://developer.mozilla.org/en-US/docs/Glossary/Truthy

File: task/02-numbers-tasks.js [hunk 9]
```diff
@@ -127,7 +131,9 @@ function getAngleBetweenVectors(x1, y1, x2, y2) {
  *     0     => 0
  */
 function getLastDigit(value) {
-    throw new Error('Not implemented');
+    let num_str=value.toString();
+    let last_dig=num_str.length-1;
+    return +num_str[last_dig];
```
Comment (vramaniuk) [inline task/02-numbers-tasks.js]: Try to solve with findigng remainder of division by 10

File: task/04-arrays-tasks.js [hunk 10]
```diff
@@ -175,7 +188,7 @@ function getHead(arr, n) {
  *    [ 'a', 'b', 'c', 'd'], 3  => [ 'b', 'c', 'd' ]
  */
 function getTail(arr, n) {
-   throw new Error('Not implemented');
+   return arr.slice(arr.length - n, arr.length);
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: just `.slice(-n)`

**Review by vramaniuk — COMMENTED** [review]
Please, try to follow these rules https://javascript.info/coding-style
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by vramaniuk** [comment]
You shoud have provided this new updated link to actual travis tests
https://travis-ci.org/github/VladSelikh/js-assignments/builds/751992214
As we can see , this PR contains redundant changes
Selection_011
These files shoudn't have been changed in this PR.
Please, revert commits that have changes concerning another tasks ( 5,6)
You must create special branch for tasks 1-4 (called task_1-4 for example) `git branch task_1-4 `  
This PR should be created from branch task_1-4, not master
Selection_013

Then you should  perform commands `git chechout master`  ;  `git checkout -b task_5`  
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by VladSelikh** [comment]
Thanks!
I will create another PR from another branch.

**Pull request closed without merging**